plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -528,14 +530,14 @@ data Context uni fun
     | NoFrame
     deriving stock (Show)
 
-toExMemory :: (Closed uni, uni `Everywhere` ExMemoryUsage) => CekValue uni fun -> ExMemory
-toExMemory = \case
-    VCon c      -> memoryUsage c
-    VDelay {}   -> 1
-    VLamAbs {}  -> 1
-    VBuiltin {} -> 1
-{-# INLINE toExMemory #-}  -- It probably gets inlined anyway, but an explicit pragma
-                           -- shouldn't hurt.
+-- See Note [ExMemoryUsage instances for non-constants].
+instance (Closed uni, uni `Everywhere` ExMemoryUsage) => ExMemoryUsage (CekValue uni fun) where
```

Comment:
I feel a bit weird about this, but I have no good reason for it.

Reply:
Why? It's the same old function, except as a class method now.

I would kinda like to forbid costing of `CekValue`s in the first place and just say that if a function accepts a term whose type is a type variable, the costing function is supposed to either ignore the `ExMemory` of that argument or to unlift the argument and only then look at the `ExMemory` of the unlifted value, but I have no idea how to express that and even if we could, I think it would be too much of a complication to worth this static guarantee.

Right now I'm just keeping the old behavior and optimizing the machinery.

Reply:
I think the reason is that before it was more explicit whereas now it feels like we could accidentally use the typeclass method on something that's not a `CekValue` and not notice. But I don't think this is really a problem.

Reply:
> now it feels like we could accidentally use the typeclass method on something that's not a `CekValue`

But that's exactly what we do! We call the costing functions over the `ExMemory` of unlifted values now. It's just that we occasionally unlift into `Opaque` for polymorphic builtins, which is a wrapper around `CekValue`, hence why we need the instance. For `Integer`, `Bool`, `SomeConstant` etc we call `memoryUsage` directly without going through `CekValue` twice (once for unlifting, once for costing). It's the whole point of this PR!

Reply:
Apparently I have no idea what the point of this PR is ;)